src/source_type.h

```diff
@@ -0,0 +1,45 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file source_type.h Type for the source of cargo. */
+
+#ifndef SOURCE_TYPE_H
+#define SOURCE_TYPE_H
+
+#include "company_type.h"
+#include "industry_type.h"
+#include "town_type.h"
+
+/** Types of cargo source and destination */
+enum class SourceType : uint8_t {
+	Industry = 0, ///< Source/destination is an industry
+	Town = 1, ///< Source/destination is a town
+	Headquarters = 2, ///< Source/destination are company headquarters
+};
+
+using SourceID = uint16_t; ///< Contains either industry ID, town ID or company ID (or Source::Invalid)
+
+/** A location from where cargo can come from (or go to). Specifically industries, towns and headquarters. */
+struct Source {
+public:
+	static constexpr SourceID Invalid = 0xFFFF; ///< Invalid/unknown index of source
```

Comment:
Do we need some kind of `static_assert` to ensure this is the same as `INVALID_INDUSTRY` etc?

Reply:
It won't be equal to `INVALID_COMPANY`, and it's specifically used for denoting that a `Source` is invalid.

I would reckon you'd have an invalid `SourceType` for that, but that's not the case. Changing that would explode the size of this PR, which I would like to avoid for now. Small steps, please.